From d3a152528a3c0a9d170d552a003774b4d098dfa9 Mon Sep 17 00:00:00 2001
From: Olivier Moysan <olivier.moysan@foss.st.com>
Date: Thu, 14 Nov 2024 09:13:58 +0100
Subject: [PATCH] iio: adc: stm32-mdf: fix channel parsing

Fix channel parsing for audio use case.
In audio filter nodes the DAI child node has to be ignored to parse only
the channel child nodes.
Moreover the channel child node has to be parsed both in audio and
analog use cases.

Change-Id: I0dc87c2b1ac1e14164cdb3744531c884a60efe2d
Signed-off-by: Olivier Moysan <olivier.moysan@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/420084
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
---
 drivers/iio/adc/stm32-mdf-adc.c | 16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

diff --git a/drivers/iio/adc/stm32-mdf-adc.c b/drivers/iio/adc/stm32-mdf-adc.c
index e2dedf5f5081..2deeb29bc24b 100644
--- a/drivers/iio/adc/stm32-mdf-adc.c
+++ b/drivers/iio/adc/stm32-mdf-adc.c
@@ -1206,13 +1206,13 @@ static int stm32_mdf_adc_chan_init_one(struct iio_dev *indio_dev, struct fwnode_
 	ch->indexed = 1;
 	ch->scan_index = idx;
 
-	if (adc->dev_data->type == STM32_MDF_IIO) {
-		ret = stm32_mdf_channel_parse_of(indio_dev, node, ch);
-		if (ret < 0) {
-			dev_err(&indio_dev->dev, "Failed to parse channel [%d]\n", idx);
-			return ret;
-		}
+	ret = stm32_mdf_channel_parse_of(indio_dev, node, ch);
+	if (ret < 0) {
+		dev_err(&indio_dev->dev, "Failed to parse channel [%d]\n", idx);
+		return ret;
+	}
 
+	if (adc->dev_data->type == STM32_MDF_IIO) {
 		ch->info_mask_separate = BIT(IIO_CHAN_INFO_RAW) | BIT(IIO_CHAN_INFO_SCALE) |
 					 BIT(IIO_CHAN_INFO_OFFSET);
 		ch->scan_type.shift = 8;
@@ -1235,6 +1235,10 @@ static int stm32_mdf_adc_chan_init(struct iio_dev *indio_dev, struct iio_chan_sp
 	int chan_idx = 0, ret;
 
 	device_for_each_child_node(indio_dev->dev.parent, child) {
+		/* Skip the child nodes with a compatible. (e.g. DAI node for audio) */
+		if (fwnode_property_present(child, "compatible"))
+			continue;
+
 		ret = stm32_mdf_adc_chan_init_one(indio_dev, child, &channels[chan_idx], chan_idx);
 		if (ret < 0) {
 			dev_err(&indio_dev->dev, "Channels [%d] init failed\n", chan_idx);
-- 
2.50.1

